# Recall Machine Learning Linear Regression

At the end of this Lesson the studen will remember the main steps to train a model:

 - Split dataset in train and test subsets
 - Standardize continuous varuables
 - Transform categorical variables to dummy
 - Train linear regression models
 - Train classification models
 - Interpret the error and accuracy metrics to validate the built models

**You have two exercises at the end of the notebook**

### Import data and libraries

In [21]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
from sklearn.linear_model import LinearRegression
from sklearn import metrics

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler

In [22]:
df = pd.read_csv('Fish.csv')
df

Species  Weight  Length1  Length2  Length3   Height   Width
0     Bream   242.0     23.2     25.4     30.0  11.5200  4.0200
1     Bream   290.0     24.0     26.3     31.2  12.4800  4.3056
2     Bream   340.0     23.9     26.5     31.1  12.3778  4.6961
3     Bream   363.0     26.3     29.0     33.5  12.7300  4.4555
4     Bream   430.0     26.5     29.0     34.0  12.4440  5.1340
..      ...     ...      ...      ...      ...      ...     ...
154   Smelt    12.2     11.5     12.2     13.4   2.0904  1.3936
155   Smelt    13.4     11.7     12.4     13.5   2.4300  1.2690
156   Smelt    12.2     12.1     13.0     13.8   2.2770  1.2558
157   Smelt    19.7     13.2     14.3     15.2   2.8728  2.0672
158   Smelt    19.9     13.8     15.0     16.2   2.9322  1.8792

[159 rows x 7 columns]

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159 entries, 0 to 158
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Species  159 non-null    object 
 1   Weight   159 non-null    float64
 2   Length1  159 non-null    float64
 3   Length2  159 non-null    float64
 4   Length3  159 non-null    float64
 5   Height   159 non-null    float64
 6   Width    159 non-null    float64
dtypes: float64(6), object(1)
memory usage: 8.8+ KB


### Species variable treatment

Species is a categorical variable, hence we need to transform it to dummies before inserting in the model

https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html


In [24]:
df.Species.value_counts()

Perch        56
Bream        35
Roach        20
Pike         17
Smelt        14
Parkki       11
Whitefish     6
Name: Species, dtype: int64

Firstly, let's reduce the categories to Perch, Bream and Others

In [25]:
def fish_species(x):
    if x == 'Perch':
        return 'Perch'
    elif x == 'Bream':
        return 'Bream'
    else:
        return 'Others'

df['fish_species'] = df['Species'].apply(fish_species)    
df

Species  Weight  Length1  Length2  Length3   Height   Width fish_species
0     Bream   242.0     23.2     25.4     30.0  11.5200  4.0200        Bream
1     Bream   290.0     24.0     26.3     31.2  12.4800  4.3056        Bream
2     Bream   340.0     23.9     26.5     31.1  12.3778  4.6961        Bream
3     Bream   363.0     26.3     29.0     33.5  12.7300  4.4555        Bream
4     Bream   430.0     26.5     29.0     34.0  12.4440  5.1340        Bream
..      ...     ...      ...      ...      ...      ...     ...          ...
154   Smelt    12.2     11.5     12.2     13.4   2.0904  1.3936       Others
155   Smelt    13.4     11.7     12.4     13.5   2.4300  1.2690       Others
156   Smelt    12.2     12.1     13.0     13.8   2.2770  1.2558       Others
157   Smelt    19.7     13.2     14.3     15.2   2.8728  2.0672       Others
158   Smelt    19.9     13.8     15.0     16.2   2.9322  1.8792       Others

[159 rows x 8 columns]

### Get dummies

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.get_dummies.html

https://stats.stackexchange.com/questions/350492/why-do-we-create-dummy-variables

https://towardsdatascience.com/what-are-dummy-variables-and-how-to-use-them-in-a-regression-model-ee43640d573e

In [26]:
df_dum = pd.get_dummies(df.fish_species)
df = df.merge(df_dum, right_index = True, left_index = True, how = 'left')


In [27]:
df.drop(['Species','fish_species'], axis = 1, inplace = True)
df.columns

Index(['Weight', 'Length1', 'Length2', 'Length3', 'Height', 'Width', 'Bream',
       'Others', 'Perch'],
      dtype='object')

### Train test split

It is mandatory to randomly divide the dataset into two. One for training the model and the test split for validate it.

If error metrics are low with the test split means that our model is robust


https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html


In [28]:
fish_train, fish_test = train_test_split(df, test_size=0.2, random_state=0)

In [29]:
print(fish_train.shape)
print(fish_test.shape)

(127, 9)
(32, 9)


### Standardize the numerical variables

Sometimes numerical variables in our dataset have very different scales, taht's to have very different values between one column and other. That can harm model accuracy.

For solve this situation, we standardize, that's to put every continuous variable centered in 0 and with standard deviation 1

We **first** standardize the training set, then the test set with the training set parameters

We do not standardize the target variable

https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html#sklearn.preprocessing.StandardScaler

https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler


https://www.askpython.com/python/examples/standardize-data-in-python#:~:text=Ways%20to%20Standardize%20Data%20in%20Python%201%201.,load_iris%20...%202%202.%20Using%20StandardScaler%20%28%29%20function



In [ ]:
scale= StandardScaler()

variables_sc = ['duration_ms', 'loudness', 'tempo']

scale_fit = scale.fit(X_train1[variables_sc])

X_train_sc = pd.DataFrame(scale.transform(X_train1[variables_sc]), columns = variables_sc)

X_test_sc = pd.DataFrame(scale.transform(X_test1[variables_sc]), columns = variables_sc)

X_train_sc.shape

X_train1.drop(variables_sc, axis = 1, inplace = True)
X_train1 = X_train1.reset_index(drop = True)
y_train1 = y_train1.reset_index(drop=True)
X_train = pd.concat([X_train1, X_train_sc], axis = 1)

X_test1.drop(variables_sc, axis = 1, inplace = True)
X_test1 = X_test1.reset_index(drop = True)
y_test1 = y_test1.reset_index(drop=True)
X_test = pd.concat([X_test1, X_test_sc], axis = 1)

In [30]:
scale= StandardScaler()
variables_sc = ['Length1', 'Length2', 'Length3', 'Height', 'Width']

X_train = fish_train[['Length1', 'Length2', 'Length3', 'Height', 'Width', 'Bream','Others', 'Perch']]
y_train  = fish_train['Weight']

X_test = fish_test[['Length1', 'Length2', 'Length3', 'Height', 'Width', 'Bream', 'Others', 'Perch']]
y_test  = fish_test['Weight']


scale_train = scale.fit(X_train[variables_sc])

X_train_sc = pd.DataFrame(scale_train.transform(X_train[variables_sc]), columns = [variables_sc])
X_train = X_train.drop(variables_sc, axis = 1) # , inplace = True
X_train = X_train.reset_index(drop = True)
X_train = pd.concat([X_train, X_train_sc], axis = 1)
X_train.columns = ['Length1', 'Length2', 'Length3', 'Height', 'Width', 'Bream','Others', 'Perch']
y_train = y_train.reset_index(drop=True)

X_test_sc = pd.DataFrame(scale_train.transform(X_test[variables_sc]), columns =[variables_sc])
X_test = X_test.drop(variables_sc, axis = 1) # , inplace = True
X_test = X_test.reset_index(drop = True)
X_test = pd.concat([X_test, X_test_sc], axis = 1)
X_test.columns = ['Length1', 'Length2', 'Length3', 'Height', 'Width', 'Bream','Others', 'Perch']
y_test = y_test.reset_index(drop=True)

## Linear Regression

https://medium.com/swlh/interpreting-linear-regression-through-statsmodels-summary-4796d359035a

In [31]:
X_train = sm.add_constant(X_train)
result = sm.OLS(y_train, X_train).fit()

print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                 Weight   R-squared:                       0.895
Model:                            OLS   Adj. R-squared:                  0.889
Method:                 Least Squares   F-statistic:                     144.5
Date:                Sat, 23 Sep 2023   Prob (F-statistic):           4.55e-55
Time:                        19:10:56   Log-Likelihood:                -774.25
No. Observations:                 127   AIC:                             1564.
Df Residuals:                     119   BIC:                             1587.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        304.3119     10.022     30.365      0.0

In [32]:
X_train['predict'] = result.predict(X_train)

X_test = sm.add_constant(X_test)
X_test['predict'] = result.predict(X_test)

In [33]:
def mean_absolute_percentage_error(y_true, y_pred):
    return np.mean(np.abs((sum(y_true) - sum(y_pred)) / sum(y_true))) * 100  

print("MAE: ", metrics.mean_absolute_error(y_train, X_train['predict']))
print("MSE: ", metrics.mean_squared_error(y_train, X_train['predict']))
print("RMSE: ", np.sqrt(metrics.mean_squared_error(y_train, X_train['predict'])))
print("MAPE: ", mean_absolute_percentage_error(y_train, X_train['predict']))
print("R2: ", metrics.r2_score(y_train, X_train['predict']))


MAE:  79.48159685784069
MSE:  11556.170674549498
RMSE:  107.49963104378311
MAPE:  6.033857813902521e-14
R2:  0.8947146195772225


In [34]:
print("MAE: ", metrics.mean_absolute_error(y_test, X_test['predict']))
print("MSE: ", metrics.mean_squared_error(y_test, X_test['predict']))
print("RMSE: ", np.sqrt(metrics.mean_squared_error(y_test, X_test['predict'])))
print("MAPE: ", mean_absolute_percentage_error(y_test, X_test['predict']))
print("R2: ", metrics.r2_score(y_test, X_test['predict']))


MAE:  99.17774652309299
MSE:  26636.577496367645
RMSE:  163.20716129008446
MAPE:  3.1590670956093923
R2:  0.8600657372046434


## Exercise 1

Response the answers in 4-5 lines each, read the links you have along this document, or in the theory notebooks, or you can also search on the internet:

 - Which type of variables do we transform into dummies? Why do we do it?

   We transform categorical variables into dummy variables to represent non-numeric data in mathematical models. This allows us to preserve the information of the original categories and avoid erroneous interpretations of ordinal relationships. The most common method is one-hot encoding, where each category becomes a binary column. Additionally, it's important to drop one category to prevent multicollinearity and the "dummy variable trap."

 - Why is so important to divide our data into train and test datasets? Which is the purpose of doing it?

    It is important to divide our data into train and test datasets to avoid overfitting. The purpose of doing it is to train the model with the train dataset and then test it with the test dataset to see if the model is robust.

 - Why do we standardize some varaiables? Which type of variables do we standardize?

    We standardize some variables to avoid having variables with very different scales, which can harm model accuracy. We standardize continuous variables.

## Exercise 2

Regarding the summary and the errors, Would you use this model to predict the weights of the fishes? Justify your answer. Comment the usefulness of the main indicators of the summary and the errors.


This model has a high R-squared and a significant F-statistic, indicating a strong relationship between predictors and fish weights. However, concerns about multicollinearity and non-normal residuals should be addressed before using it for predictions. Further validation and consideration of specific application needs are necessary.

1. R-squared and Adj. R-squared:

    -Usefulness: These metrics provide a measure of how well the independent variables explain the variation in the dependent variable. An R-squared of 0.895 indicates a strong relationship, suggesting that the model explains a significant portion of the variation in fish weights.

2. F-statistic and P-value:

    -Usefulness: The low p-value (4.55e-55) for the F-statistic suggests that the model is statistically significant as a whole.

3. Coefficients and P-values for Individual Predictors:

    -Usefulness: The coefficients and their associated p-values provide information about the strength and significance of the relationship between each predictor and the dependent variable.

4. Omnibus and Jarque-Bera Tests:

    -Usefulness: These tests check the assumptions of normality for the residuals.

5. Durbin-Watson Statistic:

    Usefulness: It tests for autocorrelation in the residuals.

6. Cond. No. (Condition Number):

    -Usefulness: It measures multicollinearity.